### PR TITLE
Add new object/array-destructuring options to typedef rule

### PR DESCRIPTION
**Pull request by shlomiassaf**
#### PR checklist

- [x] Addresses an existing issue: #1039
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### What changes did you make?
This PR adds support for opting in to `object-destructuring` and `array-destructuring` in **typedef** rule.  

Current `variable-declaration` option, if on, will fail if:
```
const [ paramA, paramB ] = [15, 'test'];
```

With this PR it wont, `variable-declaration` only fails on simple non-binding declarations.

Tests samples:
```
const { paramA, paramB } = { paramA: "test", paramB: 15 };
                        ~    [expected object-destructuring: '{ paramA, paramB }' to have a typedef]

const { paramA, paramB }: { paramA: string, paramB: number } = { paramA: "test", paramB: 15 };

const [ paramA, paramB ] = [15, 'test'];
                        ~    [expected array-destructuring: '[ paramA, paramB ]' to have a typedef]

const [ paramA3, paramB3 ]: { number: string, paramB1: string } = [15, 'test'];
```
#### Is there anything you'd like reviewers to focus on?
`variable-declaration` now has it's own test  directory.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @shlomiassaf! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: test/rules/typedef/none/test.ts.lint [hunk 1]
```diff
@@ -77,3 +77,7 @@ try {
 for (let i of [1, 2, 3]) {
 
 }
+
+const [ paramA3, paramB3 ]: { number: string, paramB1: string } = [15, 'test'];
```
Comment (nchen63) [inline test/rules/typedef/none/test.ts.lint]: add destructuring cases without typedefs. Also this typedef doesn't match

Comment (shlomiassaf) [inline test/rules/typedef/none/test.ts.lint]: @nchen63 good catch, copy/paste :)
updated.

**Comment by nchen63** [comment]
@shlomiassaf thanks!

**Comment by shlomiassaf** [comment]
@nchen63 I think now it is required in typedef-whitespace
not sure I will have the time for this...

**Pull request merged**